File: usb_camera/src/video_view.cpp
#include <ros/ros.h>
#include <image_transport/image_transport.h>
#include <cv_bridge/cv_bridge.h>
#include <sensor_msgs/Image.h>
#include <opencv2/opencv.hpp>


using namespace sensor_msgs;
using namespace cv;
using namespace std;


// Global variables
string save_folder;
bool show_flag;
bool save_flag;
char filename[128];
string encoding;
Mat img;


//--------------------------- Callback funciton ---------------------------//
void ImageCallback(const ImageConstPtr& msg)
{
    ROS_INFO("Sequence: %d", msg->header.seq);

    // Transform image from ImageConstPtr to cv::Mat
    cv_bridge::CvImagePtr cv_ptr_image;
    cv_ptr_image = cv_bridge::toCvCopy(msg, encoding);
    img = cv_ptr_image->image;

    if(show_flag)
    {
        namedWindow("iamge", 0);
        imshow("iamge", img);
        waitKey(1);
    }

    if(save_flag)
    {
        sprintf(filename,"%05d.jpg", msg->header.seq);
        std::string str(filename);
        imwrite(save_folder + str, img);
    }
}

int main(int argc,char** argv)
{
    // Initial node
    ros::init(argc,argv,"video_view");
    ros::NodeHandle nh;
    ros::NodeHandle nh_priv("~");
    ROS_INFO("Start the video_view node\n");

    // Parameter
    nh_priv.param<string>("save_folder", save_folder, ""); 
    nh_priv.param<bool>("show_flag", show_flag, true);  
    nh_priv.param<bool>("save_flag", save_flag, false); 
    nh_priv.param<string>("encoding", encoding, "bgr8");   

    // Subscribe image topic
    ros::Subscriber sub = nh.subscribe("image", 10, ImageCallback);

    // Loop and wait for callback
    ros::spin();
}

File: src/video_view.py
#! /usr/bin/env python

import os
import shutil
import rospy
from sensor_msgs.msg import Image
from sensor_msgs.msg import CameraInfo
from cv_bridge import CvBridge
import cv2

###--------------------------- Global variable ---------------------------###
bridge = CvBridge()

# parameter
save_folder = None
show_flag = None
save_flag = None
encoding = None


###--------------------------- Internal funciton ---------------------------###
def image_callback(msg):
    rospy.loginfo('Sequence: %d', msg.header.seq)

    # Transform image from ImageConstPtr to cv::Mat
    img = bridge.imgmsg_to_cv2(msg, encoding)

    if show_flag:
        cv2.namedWindow('image', 0)
        cv2.imshow('image', img)
        cv2.waitKey(1)

    if save_flag:
        filename = os.path.join(save_folder, '%06d.png' % msg.header.seq)
        cv2.imwrite(filename, img)


def video_view():
    # Initial node
    rospy.init_node('video_view', anonymous=True)
    rospy.loginfo('Start the video_view node')

    # Parameter
    global save_folder, show_flag, save_flag, encoding
    save_folder = rospy.get_param('~save_folder', None)
    show_flag = rospy.get_param('~show_flag', False)
    save_flag = rospy.get_param('~save_flag', False)
    encoding = rospy.get_param('~encoding', 'bgr8')

    # Path
    if save_flag:
        if save_folder is None:
            rospy.logerr('save_folder is None!')
            return
        if os.path.exists(save_folder):
            shutil.rmtree(save_folder)
        os.makedirs(save_folder)

    # Subscribe image topic
    rospy.Subscriber('image', Image, image_callback, queue_size=10)

    # Loop and wait for callback
    rospy.spin()


if __name__ == '__main__':

    video_view()
